implementations/rust/ockam/ockam_entity/src/entity.rs

```diff
@@ -0,0 +1,363 @@
+use crate::{
+    Contact, ContactsDb, KeyAttributes, ProfileAuth, ProfileChangeEvent, ProfileChanges,
+    ProfileContacts, ProfileEventAttributes, ProfileIdentifier, ProfileIdentity, ProfileImpl,
+    ProfileSecrets, ProfileTrait, ProfileVault, ProfileVaultAccess,
+};
+
+use crate::EntityError::ProfileNotFound;
+use ockam_core::Result;
+use ockam_vault::ockam_vault_core::{PublicKey, Secret};
+
+/// An Entity represents an identity in various authentication contexts.
+#[derive(Clone)]
+pub struct Entity<V: ProfileVault> {
+    default_profile_identifier: ProfileIdentifier,
+    profiles: Vec<ProfileImpl<V>>,
+}
+
+impl<V: ProfileVault> Entity<V> {
+    #[allow(dead_code)]
+    pub fn new(default_profile: ProfileImpl<V>) -> Self {
+        let idref = default_profile.identifier();
+        let default_profile_identifier = ProfileIdentifier::from_key_id(idref.key_id().clone());
+        let profiles = vec![default_profile];
+        Entity {
+            default_profile_identifier,
+            profiles,
+        }
+    }
+
+    fn default_profile(&self) -> Option<&ProfileImpl<V>> {
+        self.profiles
+            .iter()
+            .find(|profile| &self.default_profile_identifier == profile.identifier())
+    }
+}
+
+impl<V: ProfileVault> ProfileAuth for Entity<V> {
+    fn generate_authentication_proof(&mut self, channel_state: &[u8]) -> Result<Vec<u8>> {
+        if let Some(profile) = self.default_profile() {
+            profile.clone().generate_authentication_proof(channel_state)
+        } else {
+            Err(ProfileNotFound.into())
+        }
+    }
+
+    fn verify_authentication_proof(
+        &mut self,
+        channel_state: &[u8],
+        responder_contact_id: &ProfileIdentifier,
+        proof: &[u8],
+    ) -> Result<bool> {
+        if let Some(profile) = self.default_profile() {
+            profile
+                .clone()
+                .verify_authentication_proof(channel_state, responder_contact_id, proof)
+        } else {
+            Err(ProfileNotFound.into())
+        }
+    }
+}
+
+impl<V: ProfileVault> ProfileChanges for Entity<V> {
+    fn change_events(&self) -> &[ProfileChangeEvent] {
+        if let Some(profile) = self.default_profile() {
+            profile.change_events()
+        } else {
+            &[]
+        }
+    }
+
+    fn update_no_verification(&mut self, change_event: ProfileChangeEvent) -> Result<()> {
+        for profile in &mut self.profiles {
+            if &self.default_profile_identifier == profile.identifier() {
```

Comment:
For operations that require mutable access to the default profile, each delegating method iterates the profiles instead of using `default_profile())`